osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -171,7 +176,22 @@ public void SelectRandom()
             if (visibleGroups.Count < 1)
                 return;
 
-            BeatmapGroup group = visibleGroups[RNG.Next(visibleGroups.Count)];
+            BeatmapGroup group;
+            if (randomType == SelectionRandomType.RandomPermutation)
+            {
+                List<BeatmapGroup> notSeenGroups = visibleGroups.Except(seenGroups).ToList();

```

Comment:
Is it possible to use IEnumerable<BeatmapGroup> here instead of ToList() (replacing the below `notSeenGroups[...]` with `notSeenGroups.ElementAt(...)`. This is going to be a pretty chunky allocation with a lot of beatmaps.